[Serializer] Add a normalizer that support JsonSerializable objects

mcfedr opened the pull request:
This normalizer makes it easier to start to combine using `JsonSerializable` objects with the Symfony serializer. I have implemented it in a number of projects and #13496 shows that others are doing so as well. So it seemed like it would be useful to include it in the Serializer component.

It handles circular references in the same way as the other normalizers.

Because groups and max depth are based on property annotations it doesn't make sense to apply them here.

| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #13496 |
| License | MIT |
| Doc PR |  |

fred-ekreative committed:
Add a normalizer that support JsonSerializable objects

Handles circular references

dosten commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/serializer.xml:
```diff
@@ -26,6 +26,12 @@
         </service>
 
+        <!-- Json Normalizer -->
+        <service id="serializer.normalizer.json" class="Symfony\Component\Serializer\Normalizer\JsonNormalizer" public="false">
```
Should be `Symfony\Component\Serializer\Normalizer\JsonSerializableNormalizer` or even better: rename the class to `JsonNormalizer`

mcfedr replied:
I felt like it could be confusing to shorten it as any normalizer could be making json.

dunglas commented on src/Symfony/Component/Serializer/Normalizer/JsonSerializableNormalizer.php:
```diff
@@ -0,0 +1,42 @@
+    public function normalize($object, $format = null, array $context = [])
+    {
+        if (!$object instanceof \JsonSerializable) {
+            throw new InvalidArgumentException('The object must implement "\JsonSerializable".');
```
Use `\JsonSerializable::class` constant instead.

mcfedr replied:
Like this?

`throw new InvalidArgumentException("The object must implement " . \JsonSerializable::class);`

dunglas commented on src/Symfony/Component/Serializer/Normalizer/JsonSerializableNormalizer.php:
```diff
@@ -0,0 +1,42 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function normalize($object, $format = null, array $context = [])
```
In Symfony we still use `array()` for consistency with the old code.

mcfedr replied:
There is a massive pull-request waiting to happen then :)

dunglas commented:
I left some minor comment but :+1:. It's a good improvement to the Serializer component.

Do you think it's possible to add support to normalization groups, max depth and circular references handling (extending `AbstractObjectNormalizer`?).

mcfedr commented:
Not sure how groups would work, the interface for `JsonSerializable` is basic, it would only work if you started removing keys from whatever `jsonSerialize` returns, and if you start doing its probably easier to convert to using the `ObjectNormalizer`.

Max depth and circular should be do able.

mcfedr commented:
I've added handling of circular references, but as i updated in the description, because groups and max depth are based on property annotations I don't think it makes sense to apply them to this normalizer.

dunglas commented:
Fair enough. :+1: ping @symfony/deciders

xabbuh commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/serializer.xml:
```diff
@@ -26,6 +26,12 @@
 
+        <!-- Json Normalizer -->
+        <service id="serializer.normalizer.json" class="Symfony\Component\Serializer\Normalizer\JsonSerializableNormalizer" public="false">
+            <!-- Run after all custom serializers, but before the ObjectNormalizer -->
```
Shouldn't this be "all custom normalizers"?

xabbuh commented:
should also be added to the component's changelog

mcfedr commented:
@xabbuh Is that something I should add to the pull request?

mcfedr commented:
I am sort of assuming the tests not working is nothing to do with me, but could it be? I have run the same tests locally, with no fails...

xabbuh commented:
@mcfedr That's not your fault. Our test suite is currently broken for the `3.0` and `master` branch.

xabbuh commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/serializer.xml:
```diff
@@ -26,6 +26,12 @@
         </service>
 
+        <!-- Json Normalizer -->
+        <service id="serializer.normalizer.json" class="Symfony\Component\Serializer\Normalizer\JsonSerializableNormalizer" public="false">
```
How can we handle the case that someones uses the FrameworkBundle with an older version of the Serializer component that doesn't contain this class?

mcfedr replied:
I guess the obvious answer is to add the service in the `FrameworkBundle::registerSerializerConfiguration` method after a check to see if the class exists, something similar seems to happen with some other parts of the configuration.
Alternatively there doesnt seem to be an issue with just adding the `ObjectNormalizer` to the config when it was added

dunglas replied:
We can also add something like `"conflict": {"symfony/serializer": "<3.1"}` in the `composer.json` of FrameworkBundle.

mcfedr replied:
Could add conflict, i think thats a good idea, means we don't need any extra code for legacy reasons.
I notice that `symfony/serializer` isn't in the `require-dev` section either which it probably should be.

mcfedr replied:
Added `conflict`, also added a `require-dev` on `symfony/serializer` which was missing

xabbuh replied:
I would say that we should handle it in PHP code instead of using Composer rules here (@dunglas did a similar approach for new normalizers).

stloyd commented on src/Symfony/Component/Serializer/CHANGELOG.md:
```diff
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+3.0.1
```
Not really ;)

fabpot commented:
@nicolas-grekas Can you have a look at the composer constraint changes?

dunglas commented:
An alternative is to register this normalizer conditionally as in #17631.

mcfedr commented on src/Symfony/Bundle/FrameworkBundle/composer.json:
```diff
@@ -46,6 +46,7 @@
         "symfony/form": "~2.8|~3.0",
         "symfony/expression-language": "~2.8|~3.0",
         "symfony/process": "~2.8|~3.0",
+        "symfony/serializer": "~2.8|^3.1",
```
I have left this addition for dev requirement. It would appear that is was completely missing before.

dunglas replied:
Because of how it is registered now, you can safely use `~2.8|~3.0` like for other components.

mcfedr commented:
Updated to use the same style as @dunglas - checking for existing classes

dunglas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -914,6 +915,13 @@ private function registerSerializerConfiguration(array $config, ContainerBuilder
 
+        if (class_exists('Symfony\Component\Serializer\Normalizer\JsonSerializableNormalizer')) {
+            // Run before serializer.normalizer.object
+            $definition = $container->register('serializer.normalizer.json', JsonSerializableNormalizer::class);
```
`json_serializable`? Nitpicking but it looks more explicit to me.

mcfedr replied:
See what you mean, I've changed it

dunglas commented:
Except minor comments I left, :+1: for me.

Status: Reviewed

fabpot commented:
Thank you @mcfedr.
